### PR TITLE
[Tests]Stop the quarantine of [test_id:7164] and [test_id:6993]

### PR DESCRIPTION
**Pull request by Barakmor1**
Changes:
    
1. Kill virt-laucher process instead of creating VMI with a
mismatched disk
Currently these tests Create a Kernel Boot VMI with
a mismatched disk when doing so the controller keeps
trying to mount the bogus kernel location which making
these test flacky as the test expect the VMI to fail.
    
2. Create a running VM instead of using expensive
API calls through virtctl
    
3. Check existence of launcher pod directly once the VMI
Fail instead of adding unnecessary complex conditions.
Currently, the launcher verification uses Eventually and
Consistently scopes for objects that never change.

Fixes #

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: 

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note

```release-note
NONE
```

### CODE REVIEW

File: tests/vm_test.go
Comment (xpivarc) [inline tests/vm_test.go:1329]: We should check the Pod is still running and is not terminating.

**Review by xpivarc — CHANGES_REQUESTED** [review]

File: tests/vm_test.go
Comment (Barakmor1) [reply tests/vm_test.go]: Yes, this verification is wrong; the pod should exist anyway.

> We should check the Pod is still running and is not terminating.

After digging into it, I think we should check if the compute container is still running. :)

File: tests/vm_test.go
Comment (nunnatsa) [inline tests/vm_test.go:1334]: This is a bit strange: we have two options here, one of them with many other options. Is there a reason for that?

Can we move the `libvmi.WithRunStrategy(v1.RunStrategyOnce)` option to be a parameter of the `NewVirtualMachine` option?

File: tests/vm_test.go
Comment (nunnatsa) [inline tests/vm_test.go:1354]: Please use explicit intervals:
```suggestion
					}).WithTimeout(60*time.Second).WithPolling(5*time.Second).Should(BeTrue(), "virt-launcher should be created eventually")
```

**Comment by fossedihelm** [comment]
First commit conflicts with https://github.com/kubevirt/kubevirt/pull/12029
Putting an hold based on the dependency
/hold

File: tests/vm_test.go
Comment (nunnatsa) [inline tests/vm_test.go:1356]: can we return also error here?

**Comment by Barakmor1** [comment]
> First commit conflicts with #12029 Putting an hold based on the dependency /hold

Wouldn't Tide block this from getting in if this is the case?

File: tests/vm_test.go
Comment (Barakmor1) [reply tests/vm_test.go:1334]: The reason for the different options is that one set of options is for the VM and another for the VMI. The VMI doesn't have a runStrategy field in its spec; therefore, we need both options.

File: tests/vm_test.go
Comment (Barakmor1) [reply tests/vm_test.go:1356]: Thanks, but I prefer to keep it as it is. I think it is clear enough because the function does exactly as described in its name: it returns the Compute Container State Running and not an error. I don't think we should return an error if it doesn't exist because this is expected in one of the DescribeTable entries.

File: tests/vm_test.go
Comment (Barakmor1) [reply tests/vm_test.go]: Done.

**Comment by fossedihelm** [comment]
> > First commit conflicts with #12029 Putting an hold based on the dependency /hold
> 
> Wouldn't Tide block this from getting in if this is the case?

Yes, but there is no reason to perform all the CI continuously if we already know that there will be a new push, and hence a new CI cycle.

**Review by RamLavi — COMMENTED** [review]
nit: in the sake of leaving the floor cleaner than you came, please regroup the import list, we have these that should be grouped in the kubevirt.io/kubevirt/tests imports:
```
	"kubevirt.io/kubevirt/tests/libdv"
	"kubevirt.io/kubevirt/tests/libpod"
```

File: tests/vm_test.go
Comment (nunnatsa) [reply tests/vm_test.go:1354]: @Barakmor1 We want to avoid assertions inside helpers, if possible. We prefer returning an error and assertion for no error where calling the function.

Comment (Barakmor1) [reply tests/vm_test.go:1354]: @nunnatsa There are many cases where we use assertions inside `Eventually`/ `Consistently` scopes please see: https://github.com/kubevirt/kubevirt/pull/11848#discussion_r1590296825

Comment (nunnatsa) [reply tests/vm_test.go:1354]: 1. This is not `Eventually` nor `Consistently`.
2. We want to write new code with higher standards.

There are disadvantages of using gomega assertions within helper function. First, that make the function aware of where it's going to be called, and so it's bad design. But more than that: gomega's error logs is misleading in these cases. You won't get the context of the failure - because you won't be able to know where it was called from. If you're going to use offset (`ExpectWithOffset`), then you loose the exact error if you have several assertions. 

The golang way is to return error. Only use assertions in the test itself.

File: tests/vm_test.go
Comment (Barakmor1) [reply tests/vm_test.go:1354]: I think it’s strange to return an error if containerStateRunning doesn’t exist, as this is a valid case. I prefer to have the validation check if the field exists or not, even if it means less clear logs in case of failure. I will improve the error message in case of failure. However, please feel free to avoid approving this PR if you don’t agree.

File: tests/vm_test.go
Comment (nunnatsa) [reply tests/vm_test.go:1354]: still. you can do

```go
getComputeContainerStateRunning := func() (*k8sv1.ContainerStateRunning, error) { ... }
```

And then:
```go
Eventually(func(g Gomega){
    ...
    err, running := getComputeContainerStateRunning()
    g.Expect(err).ToNot(HaveOccurred())
    ...
}).Should(Succeed())
```

Comment (nunnatsa) [reply tests/vm_test.go:1354]: Again, not complaining about lambda, but about asserting within helper function.

Comment (Barakmor1) [reply tests/vm_test.go:1354]: ok fair enough i will update this soon

**Comment by xpivarc** [comment]
@Barakmor1 

>Kill virt-laucher process instead of creating VMI with a
mismatched disk
Currently these tests Create a Kernel Boot VMI with
a mismatched disk when doing so the controller keeps
trying to mount the bogus kernel location which making
these test flacky as the test expect the VMI to fail.

Do we know why it sometimes passes?

**Comment by Barakmor1** [comment]
> @Barakmor1
> 
> > Kill virt-laucher process instead of creating VMI with a
> > mismatched disk
> > Currently these tests Create a Kernel Boot VMI with
> > a mismatched disk when doing so the controller keeps
> > trying to mount the bogus kernel location which making
> > these test flacky as the test expect the VMI to fail.
> 
> Do we know why it sometimes passes?

Actually, I'm not sure if it sometimes passes with the current master.

**Comment by xpivarc** [comment]
> > @Barakmor1
> > > Kill virt-laucher process instead of creating VMI with a
> > > mismatched disk
> > > Currently these tests Create a Kernel Boot VMI with
> > > a mismatched disk when doing so the controller keeps
> > > trying to mount the bogus kernel location which making
> > > these test flacky as the test expect the VMI to fail.
> > 
> > 
> > Do we know why it sometimes passes?
> 
> Actually, I'm not sure if it sometimes passes with the current master.

I looked into this today and locally this passes for me. I just checked issues/PRs before I dive into this mystery. That is why I am also asking how this can behave so different...

**Comment by dhiller** [comment]
@Barakmor1 what is the status of this PR? Do you need more information?

**Comment by Barakmor1** [comment]
> @Barakmor1 what is the status of this PR? Do you need more information?

Hey Daniel,

I need to rebase and address some comments. I'll continue this work soon.

**Comment by Barakmor1** [comment]
/unhold 

@fossedihelm removing the hold label as this is no longer depends on https://github.com/kubevirt/kubevirt/pull/12029

File: tests/vm_test.go
Comment (Barakmor1) [reply tests/vm_test.go]: Done

**Comment by Barakmor1** [comment]
@nunnatsa PTAL

**Comment by brianmcarey** [comment]
Hi @Barakmor1  -  what is the status of this PR? These tests have been in quarantine for 2 months now.

**Comment by Barakmor1** [comment]
> Hi @Barakmor1 - what is the status of this PR? These tests have been in quarantine for 2 months now.

Hi Brain,

The PR is ready for your review. :)

**Comment by Barakmor1** [comment]
@xpivarc @nunnatsa Can you please have a look?

File: tests/vm_test.go
Comment (xpivarc) [inline tests/vm_test.go:1349]: Because we use boolean it will be hard to say on which point we fail. I think want we care about here is to wait for the VMI to hit the `Scheduled` phase or not?

File: tests/vm_test.go
Comment (xpivarc) [inline tests/vm_test.go:1352]: Why don't we keep this?

File: tests/vm_test.go
Comment (xpivarc) [inline tests/vm_test.go:1373]: We want to check it is true as well

File: tests/vm_test.go
Comment (Barakmor1) [reply tests/vm_test.go:1349]: I think you're suggesting using something like `Eventually(ThisVMI(vmi), 60).Should(BeInPhase(v1.Scheduled))` . I tried it, but the test became unreliable and gave errors like: `VirtualMachineInstance/testvmi-vnmz7 expected phase is 'Failed' but got 'Scheduled'`. It seems the phase might be changing too quickly, so we sometimes miss the 'Scheduled' phase.

File: tests/vm_test.go
Comment (Barakmor1) [reply tests/vm_test.go:1375]: Done

File: tests/vm_test.go
Comment (Barakmor1) [reply tests/vm_test.go:1384]: I'm not sure I understand what should be true as well. Can you please elaborate?

I changed the name 'running' to something else to avoid confusion, as this is not a boolean.

Comment (xpivarc) [reply tests/vm_test.go:1384]: >I changed the name 'running' to something else to avoid confusion, as this is not a boolean.

That will be it

File: tests/vm_test.go
Comment (xpivarc) [inline tests/vm_test.go:1384]: Can we add description `ShouldNot(BeNil(), "compute container should be running")`

File: tests/vm_test.go
Comment (xpivarc) [inline tests/vm_test.go:1383]: You can rewrite this to `Consistently(getComputeContainerStateRunning)....` and it will just work.

File: tests/vm_test.go
Comment (xpivarc) [inline tests/vm_test.go:1386]: Same as above

File: tests/vm_test.go
Comment (xpivarc) [inline tests/vm_test.go:1390]: Let's also assert the Pod as whole is `Failed`

File: tests/vm_test.go
Comment (Barakmor1) [reply tests/vm_test.go:1383]: I agree that it would be simple but please see:
https://github.com/kubevirt/kubevirt/pull/12040#discussion_r1660912853

File: tests/vm_test.go
Comment (Barakmor1) [reply tests/vm_test.go:1386]: I agree that it would be simple but please see:
https://github.com/kubevirt/kubevirt/pull/12040#discussion_r1660912853

File: tests/vm_test.go
Comment (xpivarc) [reply tests/vm_test.go:1383]: I won't insist on this one

Comment (fossedihelm) [reply tests/vm_test.go:1383]: I don't think that:
```
Consistently(getComputeContainerStateRunning).WithTimeout(10 * time.Second).WithPolling(1 * time.Second).ShouldNot(BeNil())
```
goes against https://github.com/kubevirt/kubevirt/pull/12040#discussion_r1660912853.
On the contrary, it satisfies it.
IIUIC the main issue is that we should not perform plain `Expect` inside `Eventually` or `Consistently` blocks, because the context of the `Expect` is different from the `Consistently` one.
If you want to use `Expect` you should need to pass the `g Gomega` to the function.

Consider that Gomega:
```
If multiple values are returned by the function, Eventually will 
pass the first value to the matcher and require that all others 
are zero-valued. This allows you to pass Eventuallya function 
that returns a value and an error - a common pattern in Go.
```
https://onsi.github.io/gomega/#category-2-making-codeeventuallycode-assertions-on-functions

So in this case, Lubo's suggestion is the right way to use the `Consistently`.

@nunnatsa Please chime in if this is wrong

File: tests/vm_test.go
Comment (Barakmor1) [reply tests/vm_test.go:1390]: Done

File: tests/vm_test.go
Comment (Barakmor1) [reply tests/vm_test.go]: Done

File: tests/vm_test.go
Comment (nunnatsa) [reply tests/vm_test.go:1383]: There are two reasons I don't like gomega helpers:
1. The error report is confusing. It's never clear where the error came from. If you use offset, then you don't know where exactly in the helper function you got assertion error. and if you don't use offset, then the error report points to the assertion failure, but you don't know which call failed, if you have multiple calls to the helper function.
2. I think assertion is part of the test, and the test should tell a story that you should read from stat to its end. Using helpers breaks this concept in a way. Also, the go way is to return error and check them. I think that helpers should return errors, and the test should assert these errors.

Comment (nunnatsa) [reply tests/vm_test.go:1383]: Also, hiding `Expect` within a helper, does not solves the issue with the `Expect` in eventually. It just makes it harder to debug the issue. For example, I wrote this test:

```go
var _ = Describe("Checkginkgo", func() {

	i := 0
	testingSomething := func() error {
		i++
		Expect(i).To(BeNumerically(">", 5))
		return nil
	}

	It("should count", func() {
		Eventually(testingSomething).WithTimeout(10 * time.Second).WithPolling(time.Second).Should(Succeed())
	})
})
```

Here the test assert with `Expect`, without the gomega variable, from a helper function.

If the assumption above were right, then the test should succeed after 5 seconds, but it failed immediately with
```
  [FAILED] Expected
      <int>: 1
  to be >
      <int>: 5
  In [It] at: /home/nunnatsa/GolandProjects/checkginkgo/checkginkgo_suite_test.go:21 @ 09/08/24 15:43:21.101
```

Comment (nunnatsa) [reply tests/vm_test.go:1383]: I do get the expected result when I'm doing
```go
var _ = Describe("Checkginkgo", func() {
	i := 0
	testingSomething := func(g Gomega) error {
		i++
		g.Expect(i).To(BeNumerically(">", 5))
		return nil
	}

	It("should count", func() {
		Eventually(testingSomething).
			WithTimeout(10 * time.Second).
			WithPolling(time.Second).
			Should(Succeed())
	})
})
```

Comment (fossedihelm) [reply tests/vm_test.go:1383]: I understand that and I agree. I think we are saying the same thing: "Avoid using `Expect` inside the `Eventually`/`Consistently` loop. Or if you want to it, pass the `Gomega` variable and use it".
The thing is that here I also think that it is not necessary to use the `g.Expect` since we can delegate the check to the `Consistently`.
With the example you provided I am suggesting:
```go
var _ = Describe("Checkginkgo", func() {

	i := 0
	testingSomething := func() int {
		i++
		return i
	}

	It("should count", func() {
		Eventually(testingSomething).
			WithTimeout(10 * time.Second).
			WithPolling(time.Second).
			Should(BeNumerically(">", 5))
	})
})
```

And also, using the error check property of the `Consistently`:
```go
var _ = Describe("Checkginkgo", func() {

	i := 0
	testingSomething := func() (int, error) {
		i++
		if i == 2 {
			return 0, fmt.Errorf("generic")
		}
		return i, nil
	}

	It("should count", func() {
		Eventually(testingSomething).
			WithTimeout(10 * time.Second).
			WithPolling(time.Second).
			Should(BeNumerically(">", 5))
	})
})
```
This will pass as well

**Review by xpivarc — COMMENTED** [review]
/approve

**Comment by xpivarc** [comment]
@fossedihelm PTAL

File: tests/vm_test.go
Comment (xpivarc) [reply tests/vm_test.go:1077]: @nunnatsa The helper in this case doesn't do asserts. Please read Federico's comments

Comment (Barakmor1) [reply tests/vm_test.go:1077]: Hi @fossedihelm @xpivarc @nunnatsa , I think that @fossedihelm is a nice workaround, but it could be a bit confusing. In the example:

```
Eventually(testingSomething).
	WithTimeout(10 * time.Second).
	WithPolling(time.Second).
	Should(BeNumerically(">", 5))
```
It seems like we're only checking the int, so it might be hard to tell that an error was also returned from the second output variable of `func testingSomething`. That’s why I think the current approach might make it easier to catch the error. However, if there’s strong disagreement, I’m happy to adjust it. Just let me know! :)

Comment (nunnatsa) [reply tests/vm_test.go:1077]: @Barakmor1 - it will work. Gomega will also check that the last returned value is nil, but I always prefer explicit approach. I don't like magics.

**Pull request merged**
